pkg/handlers/webhookreceiver.go

```diff
@@ -93,49 +100,91 @@ func (h *WebhookReceiverHandler) processAMReceiver(d AMReceiverData, ctx context
 		return &AMReceiverResponse{Error: err, Status: "unable to list managed notifications", Code: http.StatusInternalServerError}
 	}
 
+	// Handle each firing alert
 	for _, alert := range d.Alerts.Firing() {
-		// Ignore alerts which don't have a name
-		alertname, err := alertName(alert)
+		err = h.processAlert(alert, mnl)
 		if err != nil {
-			log.WithError(err).Error("alertname missing for alert")
-			continue
+			log.WithError(err).Error("an alert could not be successfully processed")
 		}
+	}
+	return &AMReceiverResponse{Error: nil, Status: "ok", Code: http.StatusOK}
+}
 
-		// Ignore alerts which don't have the send_managed_notification label
-		if val, ok := alert.Labels[AMLabelManagedNotification]; !ok || val == "false" {
-			log.WithField("alertname", alertname).Debug("ignoring alert with no send_managed_notification label")
-			continue
-		}
+// processAlert handles the pre-check verification and sending of a notification for a particular alert
+// and returns an error if that process completed successfully or false otherwise
+func (h *WebhookReceiverHandler) processAlert(alert template.Alert, mnl *oav1alpha1.ManagedNotificationList) error {
+	// Should this alert be handled?
+	if !isValidAlert(alert) {
+		log.WithField(LogFieldAlert, alert).Info("alert does not meet valid criteria")
+		return fmt.Errorf("alert does not meet valid criteria")
+	}
 
-		// Ignore alerts which don't have a notification defined
-		notificationName, foundNotification := alert.Labels[AMLabelTemplateName]
-		if !foundNotification {
-			log.WithField("alertname", alertname).Error("alert does not have template defined")
-			continue
-		}
+	// Can the alert be mapped to an existing notification definition?
+	notification, managedNotifications, err := getNotification(alert.Labels[AMLabelManagedNotification], mnl)
+	if err != nil {
+		log.WithError(err).WithField(LogFieldAlert, alert).Warning("an alert fired which no notification template definition exists for")
+		return err
+	}
 
-		notification, mn, err := getNotification(notificationName, mnl)
-		if err != nil {
-			log.WithError(err).WithField("alertname", alertname).Warning("an alert fired which no notification template definition exists for")
-			continue
-		}
-		log.WithFields(log.Fields{"notification": notification.Name, "alertname": alertname}).Info("Found a notification")
+	// Has a servicelog already been sent and we are within the notification's "do-not-resend" window?
+	canBeSent, err := managedNotifications.CanBeSent(notification.Name)
```

Comment:
We should have a separate card to implement/improve the CanBeSent func?
From what I am reading now, it is doing simple thing only. Not sure if it could cover our use case

Reply:
Can you please clarify what you think the current `CanBeSent` implementation is missing? I consider it a complete implementation as-is, but have I forgotten something?